Add tests for a Book functionality

dmarmul opened the pull request:
Add test cases for Book and Category entities for all layers(repository, service, controller)
Screenshots with the test coverage result
https://github.com/user-attachments/assets/82ad46c2-1d50-4f01-86fa-129b8ee82e1a
https://github.com/user-attachments/assets/901dec7d-8c88-4883-b4fc-08ecce127ac6
https://github.com/user-attachments/assets/646f1f6e-3657-4b25-9999-5d5504816b2f

dmarmul committed:
Add test cases for Book and Category entities for all layers(repository, service, controller)

dmarmul committed:
Renamed some methods

dmarmul committed:
Renamed some methods

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,230 @@
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            List<String> scriptPaths = List.of(
+                    "database/books/delete-all-books-from-books-table.sql",
+                    "database/categories/delete-all-categories-from-categories-table.sql",
+                    "database/categories/add-category-to-categories-table.sql",
+                    "database/books/add-book-to-books-table.sql",
+                    "database/categories/add-category-for-books_categories-table.sql"
```
You should delete data after each method (not before). Check it everywhere

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,230 @@
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    @BeforeAll
```
better use before each for object initialization.

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,230 @@
+    @WithMockUser(roles = {"USER"})
+    @DisplayName("Get all books")
+    void getAllBooks_Success() throws Exception {
+        BookDto secondBookDto = new BookDto();
+        secondBookDto.setId(SECOND_ID);
+        secondBookDto.setTitle(BOOK_TITLE + UNIQUE_PARAM);
+        secondBookDto.setAuthor(BOOK_AUTHOR + UNIQUE_PARAM);
+        secondBookDto.setIsbn(BOOK_ISBN + UNIQUE_PARAM);
+        secondBookDto.setPrice(BOOK_PRICE);
+        secondBookDto.setCategoryIds(categoriesId);
+        BookDto thirdBookDto = new BookDto();
+        thirdBookDto.setId(THIRD_ID);
+        thirdBookDto.setTitle(UNIQUE_PARAM + BOOK_TITLE);
+        thirdBookDto.setAuthor(UNIQUE_PARAM + BOOK_AUTHOR);
+        thirdBookDto.setIsbn(UNIQUE_PARAM + BOOK_ISBN);
+        thirdBookDto.setPrice(BOOK_PRICE);
+        thirdBookDto.setCategoryIds(categoriesId);
+        List<BookDto> expected = List.of(bookDto, secondBookDto, thirdBookDto);
+
```
Better move this logic to a private method or you can create a util class for generation data. 
I also recommend you use the construction Given - When - Then

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,230 @@
+    @Test
+    @Sql(scripts = {"classpath:database/books/delete-all-books-from-books-table.sql",
+            "classpath:database/books/add-book-to-books-table.sql",
+            "classpath:database/categories/add-category-for-books_categories-table.sql"},
```
Why do you use the add script after the test method? 
Also if you execute the same scripts for each test method you can move these scripts above the class

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+
+    @Autowired
+    private ObjectMapper objectMapper;
+    @Autowired
+    private CategoryRepository categoryRepository;
```
Let’s remove unused instances ;)

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CategoryControllerTest {
```
Apply all comments that I left for the previous class

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    @Test
+    @Sql(scripts = "classpath:database/categories/delete-all-categories-from-categories-table.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {"classpath:database/categories/delete-all-categories-from-categories-table.sql",
+            "classpath:database/categories/add-category-to-categories-table.sql"},
```
why do you execute the delete script twice?

liliia-ponomarenko commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,239 @@
+                .allMatch(i -> EqualsBuilder.reflectionEquals(expected.get(i), actualList.get(i))));
+    }
+
+    @Test
+    @WithMockUser(roles = {"USER"})
+    @DisplayName("Get category by id")
```
Are you sure that you have any category in DB?

liliia-ponomarenko commented on src/test/java/org/example/bookshop/repository/BookRepositoryTest.java:
```diff
@@ -0,0 +1,49 @@
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
+    @DisplayName("Find book by id")
+    void findById_ReturnsOneBook() {
+        Book actualBook = bookRepository.findById(BOOK_ID).orElseThrow(() ->
```
in general, you don’t need test crud repository methods ;)

liliia-ponomarenko commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,242 @@
+    @Mock
+    private BookMapper bookMapper;
+
+    @BeforeAll
```
better use BeforeEach for object initialization

liliia-ponomarenko commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    @Test
+    void findAll_ShouldReturnValidListBookDto() {
```
use the same method name pattern

liliia-ponomarenko commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,242 @@
+
+    @Test
+    void findAllByCategoryId_ShouldReturnValidListBookDtoWithoutCategoryIds() {
+        Book secondBook = new Book();
+        secondBook.setTitle(TITLE);
+        secondBook.setAuthor(AUTHOR);
+        secondBook.setIsbn(ISBN);
+
+        BookDto secondBookDto = new BookDto();
+        secondBookDto.setId(BOOK_SECOND_ID);
+        secondBookDto.setTitle(secondBook.getTitle());
+        secondBookDto.setAuthor(secondBook.getAuthor());
+        secondBookDto.setIsbn(secondBook.getIsbn());
+
+        BookDtoWithoutCategoryIds firstBookDtoWithoutCategory = new BookDtoWithoutCategoryIds();
+        firstBookDtoWithoutCategory.setId(bookDto.getId());
+        firstBookDtoWithoutCategory.setTitle(bookDto.getTitle());
+        firstBookDtoWithoutCategory.setAuthor(bookDto.getAuthor());
+        firstBookDtoWithoutCategory.setIsbn(bookDto.getIsbn());
+
+        BookDtoWithoutCategoryIds secondBookDtoWithoutCategory = new BookDtoWithoutCategoryIds();
+        secondBookDtoWithoutCategory.setId(secondBookDto.getId());
+        secondBookDtoWithoutCategory.setTitle(secondBookDto.getTitle());
+        secondBookDtoWithoutCategory.setAuthor(secondBookDto.getAuthor());
+        secondBookDtoWithoutCategory.setIsbn(secondBookDto.getIsbn());
```
better move it to private methods

liliia-ponomarenko commented on src/test/resources/database/books/delete-book-from-books-table.sql:
```diff
@@ -0,0 +1 @@
+delete from books where id = 1;
```
better remove all data and insert what you need before the test ;)

liliia-ponomarenko reviewed (changes requested):
Good job! Let’s improve your solution ;)

dmarmul committed:
Fixed mistakes.
Renamed methods.
Removed some sql scripts.
Moved some logic to a private methods.
Removed testing crud repository methods.
Used BeforeEach for object initialization and used beforeAll, beforeEach and afterEach annotations.

JJJazl commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,259 @@
+        requestDto.setPrice(BOOK_PRICE);
+        requestDto.setCategories(categories);
+
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            List<String> scriptPaths = List.of(
+                    "database/categories/add-category-to-categories-table.sql",
+                    "database/books/add-book-to-books-table.sql",
+                    "database/categories/add-category-for-books_categories-table.sql"
+            );
+            for (String scriptPath : scriptPaths) {
+                ScriptUtils.executeSqlScript(connection, new ClassPathResource(scriptPath));
+            }
+        }
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    @AfterEach
+    public void afterEach(@Autowired DataSource dataSource) throws SQLException {
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            List<String> scriptPaths = List.of(
+                    "database/categories/delete-all-categories-from-categories-table.sql",
+                    "database/books/delete-all-books-from-books-table.sql"
+            );
+            for (String scriptPath : scriptPaths) {
+                ScriptUtils.executeSqlScript(connection, new ClassPathResource(scriptPath));
+            }
+        }
+    }
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,259 @@
+import org.springframework.web.context.WebApplicationContext;
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class BookControllerTest {
```
```suggestion
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Sql(scripts = {"classpath:database/categories/add-category-to-categories-table.sql",
                    "classpath:database/books/add-book-to-books-table.sql",
                    "classpath:database/categories/add-category-for-books_categories-table.sql"},
        executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
@Sql(scripts = {"classpath:database/books-categories/delete-books-categories-relations.sql", 
                    "classpath:database/categories/delete-all-categories-from-categories-table.sql",
                    "classpath"database/books/delete-all-books-from-books-table.sql"},
        executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)       
class BookControllerTest {
```

JJJazl commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    @Test
+    @Sql(scripts = {"classpath:database/books/add-two-books-to-books-table.sql",
+            "classpath:database/categories/add-second-category-to-categories-table.sql",
+            "classpath:database/categories/add-two-category-to-books_categories-table.sql"},
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
Let's move all @Sql scripts above class, it will help to understand test methods much easier, because we have the same data and state for each method. Keep it simple :)

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CategoryControllerTest {
```
The same remarks as for the BookControllerTest

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+                .webAppContextSetup(applicationContext)
+                .apply(springSecurity())
+                .build();
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            List<String> scriptPaths = List.of(
+                    "database/categories/delete-all-categories-from-categories-table.sql",
+                    "database/books/delete-all-books-from-books-table.sql"
+            );
+            for (String scriptPath : scriptPaths) {
+                ScriptUtils.executeSqlScript(connection, new ClassPathResource(scriptPath));
+            }
+        }
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+        categoryDto.setName(CATEGORY_NAME);
+        categoryDto.setDescription(CATEGORY_DESCRIPTION);
+
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection, new ClassPathResource(
+                    "database/categories/add-category-to-categories-table.sql"));
+        }
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+        }
+    }
+
+    @AfterEach
+    public void afterEach(@Autowired DataSource dataSource) throws SQLException {
+        try (Connection connection = dataSource.getConnection()) {
+            connection.setAutoCommit(true);
+            ScriptUtils.executeSqlScript(connection, new ClassPathResource(
+                    "database/categories/delete-all-categories-from-categories-table.sql"));
+        }
+    }
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Test
+    @Sql(scripts = "classpath:database/categories/add-two-categories-to-categories-table.sql",
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    @Test
+    @Sql(scripts = {"classpath:database/books/add-two-books-to-books-table.sql",
+            "classpath:database/categories/add-second-category-to-categories-table.sql",
+            "classpath:database/categories/add-two-category-to-books_categories-table.sql"},
+            executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = "classpath:database/books/delete-all-books-from-books-table.sql",
+            executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD)
```
```suggestion
```

JJJazl commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,241 @@
+import org.springframework.data.domain.Sort;
+
+@ExtendWith(MockitoExtension.class)
+class BookServiceTest {
```
Please, add @DisplayName for all methods

JJJazl commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    @Test
+    @DisplayName("verify save() method works")
```
```suggestion
    @DisplayName("Save valid book to DB")
```

JJJazl commented on src/test/java/org/example/bookshop/service/CategoryServiceTest.java:
```diff
@@ -0,0 +1,146 @@
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class CategoryServiceTest {
```
Please, add @DisplayName for all methods

JJJazl commented on src/test/java/org/example/bookshop/service/BookServiceTest.java:
```diff
@@ -0,0 +1,241 @@
+        String expected = EXCEPTION_MESSAGE + NON_EXISTING_BOOK_ID;
+        // When
+        when(bookRepository.findById(NON_EXISTING_BOOK_ID)).thenReturn(Optional.empty());
+        Exception exception = assertThrows(EntityNotFoundException.class,
+                () -> bookService.findById(NON_EXISTING_BOOK_ID)
+        );
```
Check an exception message content in all tests that throw exception

JJJazl commented on src/test/resources/database/books/add-two-books-to-books-table.sql:
Let's have one sql script to insert all books data into books table and one to insert all categories data to categories table. Please, apply this using @Sql scripts in your controller test classes

JJJazl commented on src/test/resources/database/books/add-two-books-to-books-table.sql:
```diff
@@ -0,0 +1,4 @@
+insert into books (id, title, author, isbn, price)
+values (2, 'Book 1unique', 'Author 1unique', 'ISBN 1unique', 200);
+insert into books (id, title, author, isbn, price)
+values (3, 'uniqueBook 1', 'uniqueAuthor 1', 'uniqueISBN 1', 200);
```
You don't need to write an "insert" each time
```suggestion
insert into books (id, title, author, isbn, price)
values 
(2, 'Book 1unique', 'Author 1unique', 'ISBN 1unique', 200),
(3, 'uniqueBook 1', 'uniqueAuthor 1', 'uniqueISBN 1', 200);
```

dmarmul committed:
Used Sql scripts above class.
Added DisplayName for each method.
Reduced the number of sql scripts.

Elena-Bruyako commented on src/test/java/org/example/bookshop/controller/CategoryControllerTest.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    private List<CategoryDto> createCategoriesDto() {
+        CategoryDto secondCategoryDto = new CategoryDto();
```
It's better to create TestUtil class, create static methods to generate test data and provide constants.

Elena-Bruyako commented on src/test/java/org/example/bookshop/controller/BookControllerTest.java:
```diff
@@ -0,0 +1,217 @@
+import org.testcontainers.shaded.org.apache.commons.lang3.builder.EqualsBuilder;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@Sql(scripts = {"classpath:database/categories/delete-all-categories-from-categories-table.sql",
```
It's better to remove data AFTER_TEST_METHOD and add BEFORE_TEST_METHOD

Elena-Bruyako reviewed (changes requested):
see comments

dmarmul committed:
Added data by BEFORE_TEST_METHOD and removed by AFTER_TEST_METHOD sql annotations. Need to Insert BEFORE_TEST_CLASS sql script for deleting data that added by liquibase before tests starts.
Created TestUtil class where added constants and some static methods
